android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Node.kt

```diff
@@ -0,0 +1,316 @@
+) {
+    companion object {
+        @VisibleForTesting internal val KEY_CHILD_NODES = "node.children"
+        @VisibleForTesting internal val KEY_ROUTER = "node.router"
+        @VisibleForTesting internal val KEY_INTERACTOR = "node.interactor"
+        private const val KEY_RIB_ID = "rib.id"
+        private const val KEY_VIEW_STATE = "view.state"
+        private val requestCodeRegistry = RequestCodeRegistry(8)
+    }
+
+    init {
+        router.node = this
+        interactor.node = this
+    }
+
+    private var savedInstanceState: Bundle? = null
+    val children = CopyOnWriteArrayList<Node<*>>()
+    protected var tag: String = "${this::class.java.name}.${UUID.randomUUID()}"
+        private set
+    private var ribId: Int? = null
+    internal var view: V? = null
+        private set
+    protected var parentViewGroup: ViewGroup? = null
+
+    private var savedViewState: SparseArray<Parcelable> = SparseArray()
+
+
+    private fun generateRibId(): Int =
+        requestCodeRegistry.generateGroupId(tag)
+
+    private fun generateRequestCode(code: Int): Int =
+        requestCodeRegistry.generateRequestCode(tag, code)
+
+    private fun updateRibId(value: Int) {
+        ribId = value
+    }
+
+    fun attachToView(parentViewGroup: ViewGroup) {
+        this.parentViewGroup = parentViewGroup
+        view = createView(parentViewGroup)
+        view?.let {
+            parentViewGroup.addView(it.androidView)
+            it.androidView.restoreHierarchyState(savedViewState)
+            interactor.onViewCreated(it)
+        }
+
+        children.forEach {
+            attachChildView(it)
+        }
+    }
+
+    private fun createView(parentViewGroup: ViewGroup): V? =
+        viewFactory?.invoke(parentViewGroup)
+
+    internal fun attachChild(child: Node<*>, bundle: Bundle? = null) {
+        attachChildView(child)
+        // todo refactor so that this branching is not necessary
+        if (bundle != null) {
+            attachChildNode(child, bundle)
+        } else {
+            attachChildNode(child)
+        }
+    }
+
+    internal fun attachChildView(child: Node<*>) {
+        // parentViewGroup is guaranteed to be non-null if and only if Android view system is available
+        parentViewGroup?.let {
+            child.attachToView(
+                router.getParentViewForChild(child.forClass, view) ?: it
+            )
+        }
+    }
+
+    internal fun detachChild(child: Node<*>) {
+        detachChildNode(child)
+        detachChildView(child)
+    }
+
+    internal fun saveViewState() {
+        view?.let {
+            it.androidView.saveHierarchyState(savedViewState)
+        }
+    }
+
+    internal fun detachChildView(child: Node<*>) {
+        parentViewGroup?.let {
+            child.onDetachFromView(
+                parentViewGroup = router.getParentViewForChild(child.forClass, view) ?: it
+            )
+        }
+    }
+
+    fun onDetachFromView(parentViewGroup: ViewGroup) {
```

Comment:
We have `attachToView`, shall we consider `detachFromView` as a command?
Also, should we group all `attach` and `detach` methods together visually in the file?